base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1420,19 +1420,26 @@ fn find_orphan_chain_tips<T: BlockchainBackend>(
 ) -> Vec<BlockHash>
 {
     let mut tip_hashes = Vec::<BlockHash>::new();
+    let mut next_parent_hash = None;
+    let mut next_parent_height = 0;
     db.for_each_orphan(|pair| {
         let (_, block) = pair.unwrap();
         if (block.header.prev_hash == parent_hash) && (block.header.height == parent_height + 1) {
-            let next_parent_hash = block.hash();
-            let mut orphan_chain_tips = find_orphan_chain_tips(db, block.header.height, next_parent_hash.clone());
-            if !orphan_chain_tips.is_empty() {
-                tip_hashes.append(&mut orphan_chain_tips);
-            } else {
-                tip_hashes.push(next_parent_hash);
-            }
+            // we found a match, let save to call later
+            next_parent_hash = Some(block.hash());
+            next_parent_height = block.header.height;
```

Comment:
I think you are on to something here, but I think taking the recursive call out of here will change the logic. 

There could be multiple parent_hashes that match the above criteria, so add them to a list and then iterate over them outside of the `for_each_orphan`

Reply:
Good point, will have to update that